From 9dff0d63f5f1796ffd9dcd64e6270f0f98a9191a Mon Sep 17 00:00:00 2001
From: Liu Shixin <liushixin2@huawei.com>
Date: Fri, 14 Apr 2023 16:59:09 +0800
Subject: [PATCH 0090/1646] [Huawei] mm/dynamic_hugetlb: isolate hugepage
 without dissolve

Offering: HULK
hulk inclusion
category: feature
bugzilla: 46904

--------------------------------

The memory hotplug and memory failure will dissolve freed hugepages to
buddy system, this is not the expected behavior for dynamic hugetlb.
Skip the dissolve operation for hugepages belonging to dynamic hugetlb.
For memory hotplug, the hotplug operation is not allowed, if dhugetlb
pool existed. For memory failure, the hugepage will be discard directly.

Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 include/linux/dynamic_hugetlb.h |  6 ++++++
 mm/dynamic_hugetlb.c            | 23 ++++++++++++++++++++++-
 mm/hugetlb.c                    |  7 +++++++
 3 files changed, 35 insertions(+), 1 deletion(-)

diff --git a/include/linux/dynamic_hugetlb.h b/include/linux/dynamic_hugetlb.h
index 676a292b6d06..8757f0b6f7da 100644
--- a/include/linux/dynamic_hugetlb.h
+++ b/include/linux/dynamic_hugetlb.h
@@ -109,6 +109,7 @@ int dhugetlb_acct_memory(struct hstate *h, long delta, struct hugetlbfs_inode_in
 struct page *alloc_huge_page_from_dhugetlb_pool(struct hstate *h, struct dhugetlb_pool *hpool,
 						bool need_unreserved);
 void free_huge_page_to_dhugetlb_pool(struct page *page, bool restore_reserve);
+bool page_belong_to_dynamic_hugetlb(struct page *page);
 
 #else
 
@@ -176,6 +177,11 @@ static inline
 void free_huge_page_to_dhugetlb_pool(struct page *page, bool restore_reserve)
 {
 }
+static inline
+bool page_belong_to_dynamic_hugetlb(struct page *page)
+{
+	return false;
+}
 #endif
 
 #endif /* CONFIG_DYNAMIC_HUGETLB */
diff --git a/mm/dynamic_hugetlb.c b/mm/dynamic_hugetlb.c
index c35f8f7bfa69..425e38879a7e 100644
--- a/mm/dynamic_hugetlb.c
+++ b/mm/dynamic_hugetlb.c
@@ -451,6 +451,19 @@ static struct dhugetlb_pool *find_hpool_by_dhugetlb_pagelist(struct page *page)
 	return hpool;
 }
 
+bool page_belong_to_dynamic_hugetlb(struct page *page)
+{
+	struct dhugetlb_pool *hpool;
+
+	if (!dhugetlb_enabled)
+		return false;
+
+	hpool = find_hpool_by_dhugetlb_pagelist(page);
+	if (hpool)
+		return true;
+	return false;
+}
+
 static struct dhugetlb_pool *find_hpool_by_task(struct task_struct *tsk)
 {
 	struct mem_cgroup *memcg;
@@ -743,8 +756,15 @@ void free_huge_page_to_dhugetlb_pool(struct page *page, bool restore_reserve)
 	}
 
 	spin_lock(&hpool->lock);
+	/*
+	 * memory_failure will free the hwpoison hugepage, and then try to
+	 * dissolve it and free subpage to buddy system. Since the page in
+	 * dhugetlb_pool should not free to buudy system, we isolate the
+	 * hugepage here directly, and skip the latter dissolution.
+	 */
+	if (PageHWPoison(page))
+		goto out;
 	ClearPagePool(page);
-	set_compound_page_dtor(page, NULL_COMPOUND_DTOR);
 	if (hstate_is_gigantic(h))
 		hpages_pool = &hpool->hpages_pool[HUGE_PAGES_POOL_1G];
 	else
@@ -760,6 +780,7 @@ void free_huge_page_to_dhugetlb_pool(struct page *page, bool restore_reserve)
 	}
 	trace_dynamic_hugetlb_alloc_free(hpool, page, hpages_pool->free_huge_pages,
 					 DHUGETLB_FREE, huge_page_size(h));
+out:
 	spin_unlock(&hpool->lock);
 	put_hpool(hpool);
 }
diff --git a/mm/hugetlb.c b/mm/hugetlb.c
index c4ca70609b74..1102868e5b36 100644
--- a/mm/hugetlb.c
+++ b/mm/hugetlb.c
@@ -2033,6 +2033,13 @@ int dissolve_free_huge_page(struct page *page)
 	if (!PageHuge(page))
 		return 0;
 
+	/*
+	 * the page belong to dynamic hugetlb will be isolated as a whole
+	 * when free. See free_huge_page_to_dhugetlb_pool() for detail.
+	 */
+	if (page_belong_to_dynamic_hugetlb(page))
+		return -EBUSY;
+
 	spin_lock_irq(&hugetlb_lock);
 	if (!PageHuge(page)) {
 		rc = 0;
-- 
2.17.1

